graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/runtime/JSRealm.java

```diff
@@ -1994,13 +1994,13 @@ private void putGraalObject() {
         JSObjectUtil.putDataProperty(context, graalObject, Strings.VERSION_ECMA_SCRIPT, esVersion, flags);
         JSObjectUtil.putDataProperty(context, graalObject, Strings.IS_GRAAL_RUNTIME, JSFunction.create(this, isGraalRuntimeFunction(context)), flags);
         if (options.getUnhandledRejectionsMode() == JSContextOptions.UnhandledRejectionsTrackingMode.HANDLER) {
-            JSFunctionObject registerFunction = JSFunction.create(this, setUnhandledPromiseRejectionHandlerFunction());
+            JSFunctionObject registerFunction = JSFunction.create(this, setUnhandledPromiseRejectionHandlerFunction(context));
             JSObjectUtil.putDataProperty(context, graalObject, Strings.SET_UNHANDLED_PROMISE_REJECTION_HANDLER, registerFunction, flags);
         }
         putGlobalProperty(Strings.GRAAL, graalObject);
     }
 
-    private JSFunctionData setUnhandledPromiseRejectionHandlerFunction() {
+    private static JSFunctionData setUnhandledPromiseRejectionHandlerFunction(JSContext context) {
```

Comment:
I ran into quite the unfortunate bug here. I didn't realize how the realms worked in a shared engine environment, so when testing in my system I noticed the handler wasn't being registered with my shared-engine contexts. Fix it to pull a live reference to the realm when setting, and made the `JSFunctionData` here static.  Also added a new unit test for this case.